lib/stub_ui/lib/src/ui/window.dart

```diff
@@ -35,6 +35,71 @@ typedef PlatformMessageResponseCallback = void Function(ByteData data);
 typedef PlatformMessageCallback = void Function(
     String name, ByteData data, PlatformMessageResponseCallback callback);
 
+/// Time-related performance metrics of a frame.
+class FrameTiming {
+  const FrameTiming._(this.buildStart, this.buildFinish, this.rasterStart, this.rasterFinish);
+
+  /// The time that UI thread starts building this frame.
+  ///
+  /// {@template dart.ui.FrameTiming.epoch}
+  /// This is a raw timestamp from some epoch. The epoch in all [FrameTiming] is
+  /// the same, but it may not match [DateTime]'s epoch.
+  /// {@endtemplate}
+  final Duration buildStart;
+
+  /// The time that UI thread finishes building this frame.
+  ///
+  /// {@macro dart.ui.FrameTiming.epoch}
+  final Duration buildFinish;
+
+  /// The time that GPU thread starts rasterizing this frame.
+  ///
+  /// {@macro dart.ui.FrameTiming.epoch}
+  final Duration rasterStart;
+
+  /// The time that GPU thread finishes rasterizing this frame.
+  ///
+  /// {@macro dart.ui.FrameTiming.epoch}
+  final Duration rasterFinish;
+
+  /// The time to build the frame on the UI thread.
+  ///
+  /// {@template dart.ui.FrameTiming.fps_smoothness_milliseconds}
+  /// To ensure smooth animations of K fps, this should not exceed 1000/K milliseconds.
+  /// {@endtemplate}
+  /// {@template dart.ui.FrameTiming.fps_milliseconds}
+  /// That's about 16ms for 60fps, and 8ms for 120fps.
+  /// {@endtemplate}
+  Duration get buildDuration => buildFinish - buildStart;
+
+  /// The time to rasterize the frame on the GPU thread.
+  ///
+  /// {@macro dart.ui.FrameTiming.fps_smoothness_milliseconds}
+  /// {@macro dart.ui.FrameTiming.fps_milliseconds}
+  Duration get rasterDuration => rasterFinish - rasterStart;
+
+  /// The timespan between build start and raster finish.
+  ///
+  /// To achieve the lowest latency on a K fps display, this should not exceed
+  /// 1000/K milliseconds.
+  /// {@macro dart.ui.FrameTiming.fps_milliseconds}
+  Duration get totalSpan => rasterFinish - buildStart;
+
+  @override
+  String toString() {
+    return '''
+FrameTiming(
```

Comment:
use `'$runtimeType(...)'` in case it's subclassed

Reply:
cc @zanderso @alexmarkov 

'$runtimeType` apparently shows up near the top of the CPU profile in internal benchmarks. If the user is overriding the class they can override toString too.

Reply:
The performance of debug methods like toString doesn't matter for release builds. I'd much much rather we didn't have people confused by toStrings that lie about what they are when debugging than about the performance of toString. If it's really that bad, we should fix runtimeType and toString, not work around underlying limitations.

Reply:
Done.